src/main/java/com/spotify/autoscaler/AutoscaleJob.java

```diff
@@ -286,34 +295,53 @@ int storageConstraints(final Duration samplingDuration, int desiredNodes, int cu
         currentNodes);
     log.storageUtilization(storageUtilization);
     if (minNodesRequiredForStorage > desiredNodes) {
+      registry
+          .meter(
+              clusterMetricPrefix
+                  .tagged("what", "overridden-desired-node-count")
+                  .tagged("reason", "storage-constraint")
+                  .tagged("desired-nodes", String.valueOf(desiredNodes))
+                  .tagged("min-nodes", String.valueOf(cluster.effectiveMinNodes()))
+                  .tagged("target-nodes", String.valueOf(minNodesRequiredForStorage))
+                  .tagged("max-nodes", String.valueOf(cluster.maxNodes())))
+          .mark();
       addResizeReason(
           String.format(
-              "Storage strategy: Target node count overriden(%d -> %d).",
+              "Storage strategy: Target node count overridden(%d -> %d).",
               desiredNodes, minNodesRequiredForStorage));
     }
     return Math.max(minNodesRequiredForStorage, desiredNodes);
   }
 
-  int sizeConstraints(int desiredNodes) {
-
+  private int sizeConstraints(int desiredNodes) {
     // the desired size should be inside the autoscale boundaries
     int finalNodes =
         Math.max(cluster.effectiveMinNodes(), Math.min(cluster.maxNodes(), desiredNodes));
     if (desiredNodes != finalNodes) {
+      registry
+          .meter(
+              clusterMetricPrefix
+                  .tagged("what", "overridden-desired-node-count")
+                  .tagged("reason", "range-constraint")
+                  .tagged("desired-nodes", String.valueOf(desiredNodes))
+                  .tagged("min-nodes", String.valueOf(cluster.effectiveMinNodes()))
+                  .tagged("target-nodes", String.valueOf(finalNodes))
+                  .tagged("max-nodes", String.valueOf(cluster.maxNodes())))
```

Comment:
Hmm, do you think you could tag the metric with whether is was the min or max constraint that triggered? I realize there's the case that @borislopezaraoz mentioned today where minNodes+extra is greater than maxNodes. So that could mean you should emit two metrics in some cases.

I'd suggest using the `"Reason"` tag for this distinction, e.g. with values `"min-constraint"` and `"max-constraint"` instead of `"range-constraint"`.

Reply:
OK, this is better. But since we're in the method that reasons about `desiredNodes` vs. `cluster.effectiveMinNodes()` and `cluster.maxNodes()` (line 317), I think it would be better to replace the `if (desiredNodes != finalNodes)` on line 318 with two independent `if` blocks:
`if (desiredNodes > cluster.maxNodes())`
and
`if (desiredNodes < cluster.effectiveMinNodes())`
And emit a metric in both of them (so 0, 1 or 2 metrics can be emitted by this method).

The `addResizeReason` should only happen once, so that can be in a separate `if (desiredNodes != finalNodes)` block.